html/libndt7.js

```diff
@@ -46,13 +52,31 @@ const libndt7 = (function () {
 
       // makeurl creates the URL from |settings|.
       const makeurl = function (settings) {
-        let url = ''
-        url += (settings.insecure) ? 'ws://' : 'wss://'
-        url += settings.hostname
-        if (settings.port) {
-          url += ':' + settings.port
+        let url = new URL(settings.href)
+        url.protocol = (url.protocol === 'https:') ? 'wss:' : 'ws:'
+        url.pathname = '/ndt/v7/download'
+        let params = new URLSearchParams()
+        for (let key in settings.meta) {
+          if (settings.meta.hasOwnProperty(key)) {
+            const re = /[0-9A-Za-z._]+/
+            if (!key.match(re)) {
+              throw 'Key does not match the expected regular expression: ' + key
+            }
+            let value = settings.meta[key]
+            if (typeof value === 'number' || typeof value === 'boolean') {
+              value += ''  // force conversion to string
+            }
+            if (typeof value !== 'string' || !value.match(re)) {
```

Comment:
I think we should be okay with unicode values and that it is okay to remove the `value.match(re)` check now that we can be sure that the URLencoding will be okay in all circumstances.

Reply:
Whether or not unicode keys make sense, I leave to you. But these checks feel redundant to me.

Reply:
So, my understanding is that [strings in JavaScript are UTF-16](https://kev.inburke.com/kevin/node-js-string-encoding/). Reading [the polyfill code for URLSearchParams](https://github.com/WebReflection/url-search-params/blob/master/src/url-search-params.js), I additionally understand that it _should_ call `encodeURIComponent`. In turn, that [should serialise to UTF-8](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent).

Is my understanding correct?

I am concerned that JavaScript strings are UTF-16 while Go strings are UTF-8. So, I would like to be sure that removing the checks enforcing ASCII we _may_ end up with URLs that we cannot use in Go.

(This is probably just me being paranoid and not knowing enough about the Web.)

Reply:
As a clarification, I am not thinking about using such params to build file names. But my code in another branch is serialising such params to JSON as part of saving results. Hence my concern before applying the objective improvement that you suggest. (As a note to self, if I do that, I shall also update the spec.)

Reply:
```go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"net/url"
	"strings"
)

func main() {
	m, err := url.ParseQuery(`x=1&y=2&w=%D1%88%D0%B5%D0%BB%D0%BB%D1%8B&y=3;z`)
	if err != nil {
		log.Fatal(err)
	}
	fmt.Println(toJSON(m))
}

func toJSON(m interface{}) string {
	js, err := json.Marshal(m)
	if err != nil {
		log.Fatal(err)
	}
	return strings.Replace(string(js), ",", ", ", -1)
}
```

results in:

```
{"w":["шеллы"], "x":["1"], "y":["2", "3"], "z":[""]}
```

So it looks like UTF16 bytes are handled by Go just fine.

Reply:
Excellent, thanks! So, I'll remove these four lines and merge :-)